src/liballoc/collections/btree/set.rs

```diff
@@ -1083,24 +1149,35 @@ impl<'a, T: Ord> Iterator for Intersection<'a, T> {
     type Item = &'a T;
 
     fn next(&mut self) -> Option<&'a T> {
-        loop {
-            match Ord::cmp(self.a.peek()?, self.b.peek()?) {
-                Less => {
-                    self.a.next();
+        match &mut self.inner {
+            IntersectionInner::Stitch { a_iter, b_iter } => {
+                let mut a_next = a_iter.next()?;
+                let mut b_next = b_iter.next()?;
+                loop {
+                    match Ord::cmp(a_next, b_next) {
+                        Less => a_next = a_iter.next()?,
+                        Greater => b_next = b_iter.next()?,
+                        Equal => return Some(a_next),
+                    }
                 }
-                Equal => {
-                    self.b.next();
-                    return self.a.next();
-                }
-                Greater => {
-                    self.b.next();
+            }
+            IntersectionInner::Search { a_iter, b_set } =>  {
+                loop {
+                    let a_next = a_iter.next()?;
+                    if b_set.contains(&a_next) {
+                        return Some(a_next);
+                    }
                 }
             }
         }
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
-        (0, Some(min(self.a.len(), self.b.len())))
+        let max_size = match &self.inner {
```

Comment:
This would actually be `min_size`, right?

Reply:
Had I taken the time to look up the doc of size_hint, I would have called it `upper_bound`. It's the "min" of the input sets, and "max" of the size_hint, so I'm not a fan of `min_size' either. How about `min_len`?

Reply:
`min_len` is ok with me :+1: